{% extends "layouts/django-material-kit/base.html" %}
{% load static %}

{% block title %} Template Page {% endblock %} 

{% block body_class %} index-page {% endblock %}

<!-- Specific Page CSS goes HERE  -->
{% block stylesheets %}{% endblock stylesheets %}

{% block content %}

  {% include 'includes/django-material-kit/navigation-transparent.html' %}

  <div class="page-header min-vh-80" style="background-image: url('https://images.unsplash.com/photo-1599408278953-e10a56e5c88d?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=2490&q=80')">
    <span class="mask bg-gradient-dark opacity-6"></span>
    <div class="container">
      <div class="row">
        <div class="col-md-8 mx-auto">
          <div class="text-center">
            <h1 class="text-white">Metadata Standards</h1>
            <p class="lead mb-4 text-white opacity-8">Criteria for communication with open access databases</p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="card card-body shadow-xl mx-3 mx-md-4 mt-n6">
    <div class="container">
      <div class="section text-start">
        <!-- START about-us image & text & info -->
        <section class="my-5">
          <div class="container">
            <div class="row">
              <div class="col-md-8 col-12 my-auto">
                <h3>Standards to be used for data and metadata format and content</h3>
                <p class="pe-md-5 mb-4">
                  While several metadata standards are applicable to different project data types, recent studies<sup>1,2</sup> recommend
                  use of the Genomic Standards Consortium’s (GSC) ‘Minimum information about a marker gene sequence’ MIMARKS,
                  Minimum Information about any (x) Sequence (<a href="https://github.com/GenomicsStandardsConsortium/mixs">MIxS - metadata template applicable to sequence data</a>) and Darwin
                  Core metadata standardsThe GSC in addition to the minimum information checklists includes
                  environmental packages. We propose to use these checklists and packages for documenting markers,
                  sequences and sample sites. MIMARKS incorporates standards being developed by the Consortium for
                  the Barcode of Life (CBOL), thus the checklist can be universally applied to any marker gene, from small
                  subunit rRNA to cytochrome oxidase I (COI), to all taxa, and to studies ranging from single individuals to
                  complex communities. The metadata for sample sites will include a unique identifier, a GPS designated
                  location, and a qualitative habitat characterization. Metadata for each sample will include date and time,
                  collector, processing steps that include filters applied, primers applied, and documentation of controls.
                  Environmental contextual data for each sample obtained at the time of sample collection will include
                  measurements on water temperature, pH, dissolved oxygen, turbidity, and salinity, with possible additional
                  variables as pertinent to the specific project. Detailed procedures and methodological approaches,
                  deviations from protocols, specific equipment, and chemical reagents utilized for this project will be
                  documented.  <br>
                  <a href="https://gensc.org/" class="text-info icon-move-right">More about the Genomic Standards Consortium (GSC)
                    <i class="fas fa-arrow-right text-sm ms-1"></i>
                  </a>
                </p>
              </div>
              <div class="col-md-4 col-12 my-auto">
                <a target="_blank" href="https://github.com/Maine-eDNA/medna-metadata">
                  <img class="w-100 border-radius-lg shadow-lg" src="{% static 'assets/img/logos/logo-gsc-sq.png' %}" alt="genomic standards consortium logo">
                </a>
              </div>
              </div>
          </div>
        </section>
      <!-- END about-us image & text & info -->
      </div>
      <div class="card card-body shadow-xl mx-3 mx-md-4 mt-n6">
        <p class="pe-md-5">
          <sup>1</sup>Tedersoo, L., Ramirez, K. S., Nilsson, R. H., Kaljuvee, A., Kõljalg, U., & Abarenkov, K. (2015). Standardizing metadata and taxonomic identification in metabarcoding studies. GigaScience, 4(1), 34. https://doi.org/10.1186/s13742-015-0074-5 <br><br>
          <sup>2</sup>Yilmaz, P., Kottmann, R., Field, D., Knight, R., Cole, J. R., Amaral-Zettler, L., Gilbert, J. A., Karsch-Mizrachi, I., Johnston, A., Cochrane, G., Vaughan, R., Hunter, C., Park, J., Morrison, N., Rocca-Serra, P., Sterk, P., Arumugam, M., Bailey, M., Baumgartner, L., … Glöckner, F. O. (2011). Minimum information about a marker gene sequence (MIMARKS) and minimum information about any (x) sequence (MIxS) specifications. Nature Biotechnology, 29(5), 415–420. https://doi.org/10.1038/nbt.1823
        </p>
      </div>
    </div>
  </div>

{% endblock content %}

<!-- Specific Page JS goes HERE  -->
{% block javascripts %}{% endblock javascripts %}
