{% extends "layouts/django-material-kit/base.html" %}
{% load static %}

{% block title %} Template Page {% endblock %} 

{% block body_class %} index-page {% endblock %}

<!-- Specific Page CSS goes HERE  -->
{% block stylesheets %}{% endblock stylesheets %}

{% block content %}

  {% include 'includes/django-material-kit/navigation-transparent.html' %}

  <div class="page-header min-vh-80" style="background-image: url('https://images.unsplash.com/photo-1599408278953-e10a56e5c88d?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=2490&q=80')">
    <span class="mask bg-gradient-dark opacity-6"></span>
    <div class="container">
      <div class="row">
        <div class="col-md-8 mx-auto">
          <div class="text-center">
            <h1 class="text-white">Metadata Standards</h1>
            <p class="lead mb-4 text-white opacity-8">Criteria for communication with open access databases</p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="card card-body shadow-xl mx-3 mx-md-4 mt-n6">
    <div class="container">
      <div class="section text-center">
    <!-- START about-us image & text & info -->
    <section class="my-5">
      <div class="container">
        <div class="row">
          <div class="col-md-8 col-12 my-auto">
            <h3>About us</h3>
            <p class="pe-md-5 mb-4">
              While several metadata standards are applicable to different project data types, recent studies1,2 recommend
              use of the Genomic Standards Consortium’s (GSC) ‘Minimum information about a marker gene sequence’ MIMARKS,
              Minimum Information about any (x) Sequence (MIxS - metadata template applicable to sequence data) and Darwin
              Core metadata standardsThe GSC in addition to the minimum information checklists includes
              environmental packages. We propose to use these checklists and packages for documenting markers,
              sequences and sample sites. MIMARKS incorporates standards being developed by the Consortium for
              the Barcode of Life (CBOL), thus the checklist can be universally applied to any marker gene, from small
              subunit rRNA to cytochrome oxidase I (COI), to all taxa, and to studies ranging from single individuals to
              complex communities. The metadata for sample sites will include a unique identifier, a GPS designated
              location, and a qualitative habitat characterization. Metadata for each sample will include date and time,
              collector, processing steps that include filters applied, primers applied, and documentation of controls.
              Environmental contextual data for each sample obtained at the time of sample collection will include
              measurements on water temperature, pH, dissolved oxygen, turbidity, and salinity, with possible additional
              variables as pertinent to the specific project. Detailed procedures and methodological approaches,
              deviations from protocols, specific equipment, and chemical reagents utilized for this project will be
              documented.  <br>
              <a href="https://umaine.edu/edna/" class="text-info icon-move-right">More about Maine-eDNA
                <i class="fas fa-arrow-right text-sm ms-1"></i>
              </a>
            </p>
          </div>
          <div class="col-md-4 col-12 my-auto">
            <a target="_blank" href="https://github.com/Maine-eDNA/medna-metadata">
              <img class="w-100 border-radius-lg shadow-lg" src="{% static 'assets/img/logo-medna-metadata.png' %}" alt="medna-metadata logo">
            </a>
          </div>
          </div>
      </div>
    </section>
    <!-- END about-us image & text & info -->
      </div>
    </div>
  </div>

{% endblock content %}

<!-- Specific Page JS goes HERE  -->
{% block javascripts %}{% endblock javascripts %}
